lxd/instance/drivers/driver_qemu.go

```diff
@@ -2044,29 +2044,34 @@ func (d *qemu) setupNvram() error {
 }
 
 func (d *qemu) qemuArchConfig(arch int) (path string, bus string, err error) {
+	basePath := ""
+	if os.Getenv("LXD_EXT_QEMU_SNAP") != "" {
```

Comment:
Does the env var need to have "SNAP" in it? From LXD's perspective this is just setting the path to QEMU right?

So I'm thinking something like `LXD_QEMU_PATH` similar to what we have for firmware here https://documentation.ubuntu.com/lxd/en/latest/environment/#server-environment-variable

Also we should have the LXD snap specify the location via the env var rather than having LXD detect that and then have the `opt/external...` bit duplicated both in LXD and in the LXD snap.

Reply:
We should add the new env var to the docs too.

Reply:
>From LXD's perspective this is just setting the path to QEMU right?

partially, yes. But not that simple, we also have to adjust AppArmor profile 
```
{{- if .extQemuSnap }}
  /snap/lxd/*/opt/external-qemu/lib/**.so*        mr,
  /snap/lxd/*/opt/external-qemu/bin/qemu-system-* mrix,
mihalicyn marked this conversation as resolved.
  /snap/lxd/*/opt/external-qemu/share/**          r,
{{- end }}
```

so, just path to Qemu won't be enough. Do you think this thing makes any sense for LXD consumers who does not use snap?

Reply:
Can we have the env var set to `/snap/lxd/*/opt/external-qemu/` (or `/snap/lxd/*/external/qemu/`) and call it `LXD_QEMU_PREFIX_PATH`?

In my view its less about whether non-snap users will want to use it and more about not having the path prefix hardcoded in two repositories.

Reply:
> In my view its less about whether non-snap users will want to use it and more about not having the path prefix hardcoded in two repositories.

got it, then I guess we can do:

- variable with name `SNAP_EXT_QEMU_PREFIX` and value like `external/qemu/`.
- AppArmor profile will be like:
```
{{- if .extQemuSnap }}
  /snap/lxd/*/{{ .snapExtQemuPrefix }}/lib/**.so*        mr,
  /snap/lxd/*/{{ .snapExtQemuPrefix }}/bin/qemu-system-* mrix,
  /snap/lxd/*/{{ .snapExtQemuPrefix }}/share/**          r,
{{- end }}
```

and path to Qemu binary calculated as `filepath.Join(os.Getenv("SNAP"), os.Getenv("SNAP_EXT_QEMU_PREFIX"), "bin/")`

Sounds good enough?

Reply:
> Can we just have `LXD_QEMU_PREFIX` (this follows similar naming format as other env vars) and then use qemu binary derived as:

In AppArmor profile we need something like `/snap/lxd/*/external/qemu`, while when we calculate path to Qemu, we need a real path prefix, like `/snap/lxd/12345/external/qemu`.

And this `*` is not a random choice, as we also have a symlink `current` in `/snap/lxd/`.

Reply:
Ah OK I see - so we need to allow access to `current` and not just the current revision?

In that case lets use `SNAP_QEMU_PREFIX` and set to `external/qemu`.

Reply:
`current` should not be visible to Apparmor at all as it always work on deref'ed paths (except when creating links but that's not related).